src/main/java/com/microsoft/azure/relay/TokenRenewer.java

```diff
@@ -61,23 +68,38 @@ void onRenewTimer() {
 	}
 
 	private void scheduleRenewTimer(SecurityToken token) {
-		this.renewTimer = new Timer();
-		Duration interval = Duration.between(Instant.now(), token.getExpiresAtUtc());
-		if (interval.isNegative()) {
-			RelayLogger.logEvent("tokenRenewNegativeDuration", this.listener);
-			return;
-		}
+	    if (this.renewTimer == null) {
+	        this.renewTimer = new Timer();    
+	    }
+	    
+	    if (scheduleLock.tryLock()) {
+	        try {
+	            Duration interval = Duration.between(Instant.now(), token.getExpiresAtUtc());
+	            if (interval.isNegative()) {
+	                RelayLogger.logEvent("tokenRenewNegativeDuration", this.listener);
+	                return;
+	            }
 
-		// TokenProvider won't return a token which is within 5min of expiring so we don't have to pad here.
-		interval = interval.compareTo(RelayConstants.CLIENT_MINIMUM_TOKEN_REFRESH_INTERVAL) < 0 ? 
-				RelayConstants.CLIENT_MINIMUM_TOKEN_REFRESH_INTERVAL : interval;
+	            // TokenProvider won't return a token which is within 5min of expiring so we don't have to pad here.
+	            interval = interval.compareTo(RelayConstants.CLIENT_MINIMUM_TOKEN_REFRESH_INTERVAL) < 0 ? 
+	                    RelayConstants.CLIENT_MINIMUM_TOKEN_REFRESH_INTERVAL : interval;
 
-		RelayLogger.logEvent("tokenRenewScheduled", this.listener, interval.toString());
-		this.renewTimer.schedule(new TimerTask() {
-			@Override
-			public void run() {
-				onRenewTimer();
-			}
-		}, interval.toMillis());
+	            if (this.renewTimerTask != null) {
+	                this.renewTimerTask.cancel();
+	            }
+                this.renewTimerTask = new TimerTask() {
+                    @Override
+                    public void run() {
+                        onRenewTimer();
+                    }
+                };
```

Comment:
Why must you cancel the old and create again? Would creating it only the first time work?
``` JAVA
             if (this.renewTimerTask == null) {
                this.renewTimerTask = new TimerTask() {
                    @Override
                    public void run() {
                        onRenewTimer();
                    }
                };
             }
```

Reply:
TimerTask cannot be reused, we have to create a new object each time. Cancel is in case if we got a new token and renewed before the timer, we probably want to avoid the task from executing since we already have another one scheduled

---
In reply to: [338835245](https://github.com/Azure/azure-relay-java/pull/54#discussion_r338835245) [](ancestors = 338835245)